spring-integration-test/src/main/java/org/springframework/integration/test/context/MockIntegrationContext.java

```diff
@@ -125,14 +126,41 @@ public void instead(String consumerEndpointId, MessageHandler mockMessageHandler
 	}
 
 	public void instead(String consumerEndpointId, MessageHandler mockMessageHandler, boolean autoStartup) {
-		instead(consumerEndpointId, mockMessageHandler, IntegrationConsumer.class, "handler", autoStartup);
+		Object endpoint = this.beanFactory.getBean(consumerEndpointId, IntegrationConsumer.class);
+		if (autoStartup && endpoint instanceof Lifecycle) {
+			((Lifecycle) endpoint).stop();
+		}
+		DirectFieldAccessor directFieldAccessor = new DirectFieldAccessor(endpoint);
+		Object targetMessageHandler = directFieldAccessor.getPropertyValue("handler");
+		this.beans.put(consumerEndpointId, targetMessageHandler);
+
 		if (mockMessageHandler instanceof MessageProducer) {
-			Object targetMessageHandler = this.beans.get(consumerEndpointId);
 			if (targetMessageHandler instanceof MessageProducer) {
 				MessageChannel outputChannel = TestUtils.getPropertyValue(targetMessageHandler, "outputChannel",
 						MessageChannel.class);
 				((MessageProducer) mockMessageHandler).setOutputChannel(outputChannel);
 			}
+			else {
+				if (mockMessageHandler instanceof MockMessageHandler) {
+					if (!TestUtils.getPropertyValue(mockMessageHandler, "hasReplies", Boolean.class)) {
+						mockMessageHandler = mockMessageHandler::handleMessage;
```

Comment:
😄 Yep! I expected such a question. 
I guess what confuses here is that we assign lambda do the same variable.

The idea is here like do not replace raw `MH` with our full reply supporting `MockMessageHandler` do not not produce unexpected behavior in the target endpoint.

In other words: 
- Our `MockMessageHandler` is `MessageProducer`, too;
- The `mockMessageHandler::handleMessage` produces just `MessageHandler`;
- I reuse replace a variable with this new instance.

Without a lambda we have an inline `MessageHandler` imple which just delegates to the provided `mockMessageHandler`.

Although I think that is overhead. we really don't need to worry about the target type from the endpoint perspective - produce or not reply is not its responsibility.
If our `MockMessageHandler` is without replies we are good to replace the raw `MH` with it and don't worry about aftermath.

Will remove this expression in the next commit.